pkg/component/pushed_component.go

```diff
@@ -0,0 +1,276 @@
+}
+
+func (d devfileComponent) GetAnnotations() map[string]string {
+	return d.d.Annotations
+}
+
+func (d devfileComponent) GetName() string {
+	return d.d.Name
+}
+
+func (d devfileComponent) GetType() (string, error) {
+	return getType(d)
+}
+
+func (d devfileComponent) GetSource() (string, string, error) {
+	return getSource(d)
+}
+
+type noSourceError struct {
+	msg string
+}
+
+func (n noSourceError) Error() string {
+	return n.msg
+}
+
+func getSource(component provider) (string, string, error) {
+	annotations := component.GetAnnotations()
+	if sourceType, ok := annotations[ComponentSourceTypeAnnotation]; ok {
+		if !validateSourceType(sourceType) {
+			return "", "", fmt.Errorf("unsupported component source type %s", sourceType)
+		}
+		var sourcePath string
+		if sourceType == string(config.GIT) {
+			sourcePath = annotations[componentSourceURLAnnotation]
+		}
+
+		klog.V(4).Infof("Source for component %s is %s (%s)", component.GetName(), sourcePath, sourceType)
+		return sourceType, sourcePath, nil
+	}
+	return "", "", noSourceError{msg: fmt.Sprintf("%s component doesn't provide a source type annotation", component.GetName())}
+}
+
+func getType(component provider) (string, error) {
+	if componentType, ok := component.GetLabels()[componentlabels.ComponentTypeLabel]; ok {
+		return componentType, nil
+	}
+	return "", fmt.Errorf("%s component doesn't provide a type label", component.GetName())
+}
+
+func GetPushedComponents(c *occlient.Client, applicationName string) (map[string]PushedComponent, error) {
+	applicationSelector := fmt.Sprintf("%s=%s", applabels.ApplicationLabel, applicationName)
+	dcList, err := c.GetDeploymentConfigsFromSelector(applicationSelector)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			dList, err := c.GetKubeClient().ListDeployments(applicationSelector)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to list components")
+			}
+			res := make(map[string]PushedComponent, len(dList.Items))
+			for _, d := range dList.Items {
+				comp, err := initPushComponent(applicationName, &devfileComponent{d: d}, c)
+				if err != nil {
+					return nil, err
+				}
+				res[comp.GetName()] = comp
+			}
+			return res, nil
+		}
+		return nil, err
+	}
+	res := make(map[string]PushedComponent, len(dcList))
+	for _, dc := range dcList {
+		comp, err := initPushComponent(applicationName, &s2iComponent{dc: dc}, c)
+		if err != nil {
+			return nil, err
+		}
+		res[comp.GetName()] = comp
+	}
+	return res, nil
+}
+
+func initPushComponent(applicationName string, p provider, c *occlient.Client) (PushedComponent, error) {
+	comp := &defaultPushedComponent{
+		application: applicationName,
+		provider:    p,
+	}
+	if err := comp.retrieveURLsAndRoutes(c); err != nil {
+		return nil, err
+	}
+	return comp, nil
+}
+
+// GetPushedComponent returns an abstraction over the cluster representation of the component
+func GetPushedComponent(c *occlient.Client, componentName, applicationName string) (PushedComponent, error) {
```

Comment:
Please add unit tests for this and some of the above functions

Reply:
These should already be covered by the existing tests, I think as this function is only meant to hide where the component data comes from. The external behavior should already be tested.